django/db/models/query_utils.py

```diff
@@ -150,6 +150,25 @@ def deconstruct(self):
             kwargs["_negated"] = True
         return path, args, kwargs
 
+    @property
+    def referenced_base_fields(self):
+        """
+        Retrieve all base fields referenced directly or through F expressions excluding
+        any fields referenced through joins.
+        """
+        # Avoid circular imports
+        from django.db.models import F
+
+        f_references = set(expr.name for expr in self.flatten() if isinstance(expr, F))
+        q_references = set(
+            child[0].split(LOOKUP_SEP)[0]
+            for expr in self.flatten()
+            if isinstance(expr, Q)
+            for child in expr.children
+            if isinstance(child, tuple)
+        )
```

Comment:
We can use `{x for x in ...}` instead of `set(x for x in ...)`:
```suggestion
        f_references = {expr.name for expr in self.flatten() if isinstance(expr, F)}
        q_references = {
            child[0].split(LOOKUP_SEP)[0]
            for expr in self.flatten()
            if isinstance(expr, Q)
            for child in expr.children
            if isinstance(child, tuple)
        }
```

Reply:
@ngnpope thanks for that I forgot sets had comprehension syntax… I might've been thinking of tuples 😅